### PR TITLE
CHE-2832: Add servers to Agent description and propagate them to Workspace Runtime

### PR DESCRIPTION
**Pull request by tolusha**
### What does this PR do?

Adds servers to Agent description and propagates them to Workspace Runtime
### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/2832

### CODE REVIEW

**Comment by tolusha** [comment]
@vparfonov @garagatyi 
pls review

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/MachineInstanceProvider.java [hunk 1]
```diff
@@ -51,8 +56,8 @@ Instance startService(String namespace,
                           boolean isDev,
                           String networkName,
                           CheServiceImpl service,
-                          LineConsumer machineLogger)
-            throws ServerException;
+                          LineConsumer machineLogger,
+                          List<? extends ServerConf> providedServers) throws ServerException;
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/MachineInstanceProvider.java]: This class works with kinda docker terminology. It knows nothing about serversConf and other Che model staf. So it is better to move this code into CheEnvironmentEngine that should apply serverConf to CheServiceImpl to describe docker container. Then server conf can be recovered from docker container labels/exposed ports. You can find that logick in DockerInstanceRuntimeInfo class

**Review by garagatyi — CHANGES_REQUESTED** [review]

**Comment by tolusha** [comment]
@garagatyi 
updated

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java [hunk 2]
```diff
@@ -376,6 +384,7 @@ public Instance startMachine(String workspaceId,
             try {
                 InstanceProvider provider = machineInstanceProviders.getProvider(machineConfig.getType());
                 machine.setId(generateMachineId());
+                addAgentsProvidedServers(machine, agents);
```
Comment (garagatyi) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java:387]: Looks like it is not called on start of the whole environment

**Review by garagatyi — CHANGES_REQUESTED** [review]

**Comment by tolusha** [comment]
@garagatyi 
updated

**Comment by slemeur** [comment]
nice!

**Review by garagatyi — APPROVED** [review]
Good job!

**Pull request merged**